### DISCUSSION TITLE
Trigger Github Pipelines On Argo Rollout Events

### DISCUSSION THREAD

**Discussion by wesleyhaws** [argoproj/argo-rollouts#2981] [Q&A] [answered]
I'm trying to figure out how to use notifications to trigger a github pipeline to run a suite of tests against a deployed resource. So I want to run a github pipeline when the deployment was a success. I have looked at the documentation here: 

* https://argo-rollouts.readthedocs.io/en/stable/generated/notification-services/github/
* https://argo-rollouts.readthedocs.io/en/stable/generated/notification-services/webhook/

There are key details missing from these docs that prevent me from moving forward with any of these ideas. The first one (https://argo-rollouts.readthedocs.io/en/stable/generated/notification-services/github/) seems like its talking about setting up a github webhook. However, webhooks in github originate from github and call out to other things. So this doesn't seem like a viable choice.

The other one (https://argo-rollouts.readthedocs.io/en/stable/generated/notification-services/webhook/) talks about using a github token. However, what about using a github app to generate a temporary token to then use that in a `Authorization: Bearer <token>` header call. It seems like this might actually be talking about a PAT (Personal Access Token) from github, which is something I want to avoid as I don't want this tied to an individual.

I didn't see any documentation talking about generating a "Custom Notification".

What are my options here as I looked like what I want to do might not be possible? Is this an impossible idea? Is it possible? Any direction would be appreciated. Thanks!

**Comment by kostis-codefresh** [comment]
The notifications you are looking at are for just sending a message such as "deployment has finished" in a fire-and-forget manner

If you want to run tests that actually decide if a rollout needs to be promoted or not you need to use the analysis subsystem

- https://argo-rollouts.readthedocs.io/en/stable/analysis/job/
- https://argo-rollouts.readthedocs.io/en/stable/analysis/web/

Can you explain a bit your use case?

Regarding notifications in general you are right that documentation can be improved. We know about this already. Please upvote https://github.com/argoproj/argo-rollouts/issues/2945

**Reply by wesleyhaws** [reply]
My original thought here would be to say when service A was successfully deployed trigger a github pipeline or circleci pipeline related to service A. When service B has successfully finished trigger a github or circle ci pipeline for service B. I want to do all of this using a github app if possible, but from reading the docs doesn't sound like is a thing. So I am trying to test out a basic webhook using a personal access token, which i would absolutely love to avoid ever having to use.

We could completely redesign our deployment process to migrate everything into argo workflows, rollouts, and CD but would love to avoid a massive lift over. As this would require generating new dockers, repositories, and other related actions for something that should be pretty simple. So I'm looking at the simple case for now before diving further in.

**Reply by kostis-codefresh** [reply]
The simplest solution if you just want to order things is to setup Rollouts with Argo CD and then use [Argo CD syncwaves](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/) to decide which Rollout happens first. Another alternative would be to put everything in an ApplicationSet and then use [progressive sync](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Progressive-Syncs/) (currently in Alpha).

The "proper" way to do ordering would be using Argo Workflows and Argo Events. Have you seen [Argo Events](https://argoproj.github.io/argo-events/concepts/event_source/)?

You can "abuse" notifications as a message/order mechanism  and a lot of people do that, but I personally think it is brittle and cumbersome.

**Comment by wesleyhaws** [comment]
Okay I have been spending over a week now scouring argocd docs, installing stuff, and overall trying to come up with ideas on how to make this, arguably simple, setup work. I found https://argo-cd.readthedocs.io/en/stable/user-guide/resource_hooks/ I built out a job, but uh-oh another hiccup. After spending over a week fighting through errors, setup problems, and finally getting this job deployed I realized something horrible.... This is generic isn't it? This is only a generic set of sync notifications for ANY sync operation in argocd. If that's the case oh crap! I lost yet another week on argocd. 

Here are my requirements:
1. This MUST be an individual notification for an individual app. This shouldn't just be a generic notification if anything gets synced in argocd. We have hundreds of applications so this isn't going to work.
2. This must allow execution of custom code for use in custom APIs to send targeted information about the last syncronized app.
3. Jobs are not created equally. A synchronized event is triggered by uploading to a repo, thats fine. However, the callback event that must take place will always be dependent on the thing that uploaded the file. So unique ids will always match to the uploader (hence not all jobs are created equal). So the jobs will only ever be triggered once and then destroyed.
4. The whole purpose of this is to be async. There shouldn't be some polling step happening somewhere eating up resources polling for hundreds of deploys. This is how it has already been setup and just like I said, it's eating resources. I want to move away from this.

So you can see why I went with a Job here. It's perfect, it will execute once with that specific set of data, and then remove itself. Jobs can run custom code that I can use with my API, and the job can be originally setup with that unique id so it knows what to call back to. However, I find myself in another, what I fear to be, week long scavanger hunt to get this done.

Any links, examples, forums, posts, etc. that can be found would be helpful. However, it seems like I have to install yet another service here "events"? Uh.... idk....

**Reply by wesleyhaws** [reply]
Parts of my journey documented across these:
https://github.com/argoproj/argo-workflows/discussions/11666
https://app.slack.com/client/T08PSQ7BQ/C01QW9QSSSK/thread/C01QW9QSSSK-1693416866.758019
https://github.com/argoproj/argo-workflows/discussions/11680
https://cloud-native.slack.com/archives/C01TSERG0KZ/p1693589355479359

This has been going on for some time now...

**Comment by wesleyhaws** [comment, marked as answer]
Ended up going with the async approach. Here is how someone else can accomplish this: https://stackoverflow.com/questions/77025678/argocd-callback-to-pipeline
